File: benchmark.py
# benchmark.py
import os
import csv
import time
import tracemalloc
import sys
from pathlib import Path
from typing import Dict, Any, List, Optional

import numpy as np

from common import load_config, make_path, SCRIPT_DIR

# ------------------------------------------------------------------
# Library detection (driven by config.libraries)
# ------------------------------------------------------------------

LIB_MODULES: Dict[str, Any] = {}


def setup_libraries(libraries: List[str]) -> None:
    """
    Try to import only the libraries requested in the YAML config.
    We store successfully imported modules in LIB_MODULES keyed by
    their config name, e.g. "chen-signatures" -> chen module.
    """
    global LIB_MODULES
    LIB_MODULES = {}

    for lib in libraries:
        if lib == "chen-signatures":
            try:
                import chen  # type: ignore
                LIB_MODULES["chen-signatures"] = chen
                print("[ok] chen-signatures available")
            except ImportError as e:
                print(f"Warning: chen-signatures import failed: {e}", file=sys.stderr)
        elif lib == "pysiglib":
            try:
                import pysiglib  # type: ignore
                LIB_MODULES["pysiglib"] = pysiglib
                print("[ok] pysiglib available")
            except ImportError as e:
                print(f"Warning: pysiglib import failed: {e}", file=sys.stderr)
        elif lib == "iisignature":
            try:
                import iisignature  # type: ignore
                LIB_MODULES["iisignature"] = iisignature
                print("[ok] iisignature available")
            except ImportError as e:
                print(f"Warning: iisignature import failed: {e}", file=sys.stderr)
        else:
            print(f"Warning: unknown library in config.libraries: {lib}", file=sys.stderr)


# ------------------------------------------------------------------
# Benchmarking helpers
# ------------------------------------------------------------------

def time_and_peak_memory(func, repeats: int = 5):
    """
    Run func() multiple times, return best time and associated peak memory.
    """
    best_time = float("inf")
    best_peak = 0

    for _ in range(repeats):
        tracemalloc.start()
        start = time.perf_counter()
        func()
        end = time.perf_counter()
        current, peak = tracemalloc.get_traced_memory()
        tracemalloc.stop()

        dur = end - start
        if dur < best_time:
            best_time = dur
            best_peak = peak

    return best_time, best_peak


# ------------------------------------------------------------------
# Per-library benchmarks
# ------------------------------------------------------------------

def bench_chen_signatures(
    d: int,
    m: int,
    N: int,
    path_kind: str,
    operation: str,
    repeats: int,
) -> Optional[Dict[str, Any]]:
    """
    Benchmarks chen-signatures for:
    - signature: chen.sig(path, m)
    - logsignature: basis = chen.prepare_logsig(d, m); chen.logsig(path, basis)
    - sigdiff:   gradient of sum(sig) via chen.torch + PyTorch
    """
    if "chen-signatures" not in LIB_MODULES:
        return None

    chen = LIB_MODULES["chen-signatures"]

    # ---------- signature ----------
    if operation == "signature":
        try:
            path = make_path(d, N, path_kind)
            path = np.ascontiguousarray(path, dtype=np.float64)

            func = lambda: chen.sig(path, m)

            # Warmup
            _ = func()

            t_sec, peak_bytes = time_and_peak_memory(func, repeats=repeats)
            t_ms = t_sec * 1000.0
            alloc_kib = peak_bytes / 1024.0

            return {
                "N": N,
                "d": d,
                "m": m,
                "path_kind": path_kind,
                "operation": operation,
                "language": "python",
                "library": "chen-signatures",
                "method": "sig",
                "path_type": "ndarray",
                "t_ms": t_ms,
                "alloc_KiB": alloc_kib,
            }
        except Exception as e:
            print(f"chen-signatures signature benchmark failed: {e}", file=sys.stderr)
            return None

    # ---------- logsignature ----------
    elif operation == "logsignature":
        # Only run if chen exposes logsig + prepare_logsig
        if not (hasattr(chen, "logsig") and hasattr(chen, "prepare_logsig")):
            print("chen-signatures: logsig/prepare_logsig not found, skipping logsignature", file=sys.stderr)
            return None

        try:
            path = make_path(d, N, path_kind)
            path = np.ascontiguousarray(path, dtype=np.float64)

            # Precompute basis outside the timed section
            basis = chen.prepare_logsig(d, m)

            func = lambda: chen.logsig(path, basis)

            # Warmup
            _ = func()

            t_sec, peak_bytes = time_and_peak_memory(func, repeats=repeats)
            t_ms = t_sec * 1000.0
            alloc_kib = peak_bytes / 1024.0

            return {
                "N": N,
                "d": d,
                "m": m,
                "path_kind": path_kind,
                "operation": operation,
                "language": "python",
                "library": "chen-signatures",
                "method": "logsig(prepared)",
                "path_type": "ndarray",
                "t_ms": t_ms,
                "alloc_KiB": alloc_kib,
            }
        except Exception as e:
            print(f"chen-signatures logsignature benchmark failed: {e}", file=sys.stderr)
            return None

    # ---------- sigdiff (autodiff) ----------
    elif operation == "sigdiff":
        # Gradient speed via PyTorch
        try:
            # Import chen.torch first (juliacall), then torch
            from chen.torch import sig_torch  # type: ignore
            import torch  # type: ignore

            path_np = make_path(d, N, path_kind)
            path_np = np.ascontiguousarray(path_np, dtype=np.float64)

            def single_run():
                path_t = torch.tensor(
                    path_np,
                    dtype=torch.float64,
                    requires_grad=True,
                )
                sig = sig_torch(path_t, m)
                loss = sig.sum()
                loss.backward()

            # Warmup a few times (compilation, Enzyme rules, etc.)
            for _ in range(3):
                single_run()

            t_sec, peak_bytes = time_and_peak_memory(single_run, repeats=repeats)
            t_ms = t_sec * 1000.0
            alloc_kib = peak_bytes / 1024.0

            return {
                "N": N,
                "d": d,
                "m": m,
                "path_kind": path_kind,
                "operation": operation,
                "language": "python",
                "library": "chen-signatures",
                "method": "sigdiff",
                "path_type": "torch",
                "t_ms": t_ms,
                "alloc_KiB": alloc_kib,
            }
        except Exception as e:
            print(f"chen-signatures sigdiff benchmark failed: {e}", file=sys.stderr)
            return None

    return None

def bench_pysiglib(
    d: int,
    m: int,
    N: int,
    path_kind: str,
    operation: str,
    repeats: int,
) -> Optional[Dict[str, Any]]:
    """
    Benchmarks pysiglib for:
    - signature: pysiglib.signature(path, degree=m)
    - sigdiff:   signature + sig_backprop (grad of sum)
    No logsig support.
    """
    if "pysiglib" not in LIB_MODULES:
        return None

    pysiglib = LIB_MODULES["pysiglib"]

    if operation == "signature":
        try:
            path = make_path(d, N, path_kind)
            path = np.ascontiguousarray(path, dtype=np.float64)

            def func():
                pysiglib.signature(path, degree=m)

            # Warmup
            _ = pysiglib.signature(path, degree=m)

            t_sec, peak_bytes = time_and_peak_memory(func, repeats=repeats)
            t_ms = t_sec * 1000.0
            alloc_kib = peak_bytes / 1024.0

            return {
                "N": N,
                "d": d,
                "m": m,
                "path_kind": path_kind,
                "operation": operation,
                "language": "python",
                "library": "pysiglib",
                "method": "signature",
                "path_type": "ndarray",
                "t_ms": t_ms,
                "alloc_KiB": alloc_kib,
            }
        except Exception as e:
            print(f"pysiglib signature benchmark failed: {e}", file=sys.stderr)
            return None

    elif operation == "sigdiff":
        # signature + sig_backprop gradient
        try:
            from pysiglib import signature, sig_backprop  # type: ignore

            path = make_path(d, N, path_kind)
            path = np.ascontiguousarray(path, dtype=np.float64)

            def single_run():
                sig = signature(path, degree=m)
                sig_derivs = np.ones_like(sig)
                _ = sig_backprop(path, sig, sig_derivs, m)

            # Warmup
            for _ in range(3):
                single_run()

            t_sec, peak_bytes = time_and_peak_memory(single_run, repeats=repeats)
            t_ms = t_sec * 1000.0
            alloc_kib = peak_bytes / 1024.0

            return {
                "N": N,
                "d": d,
                "m": m,
                "path_kind": path_kind,
                "operation": operation,
                "language": "python",
                "library": "pysiglib",
                "method": "sigdiff",
                "path_type": "ndarray",
                "t_ms": t_ms,
                "alloc_KiB": alloc_kib,
            }
        except Exception as e:
            print(f"pysiglib sigdiff benchmark failed: {e}", file=sys.stderr)
            return None

    # logsig not supported
    return None


def bench_iisignature(
    d: int,
    m: int,
    N: int,
    path_kind: str,
    operation: str,
    logsig_method: str,
    repeats: int,
) -> Optional[Dict[str, Any]]:
    """
    Benchmarks iisignature for:
    - signature: iisignature.sig(path, m)
    - logsignature: iisignature.logsig(path, basis, method)
    No sigdiff support.
    """
    if "iisignature" not in LIB_MODULES:
        return None

    iisignature = LIB_MODULES["iisignature"]

    if operation == "signature":
        try:
            path = make_path(d, N, path_kind)
            path = np.ascontiguousarray(path, dtype=np.float64)

            func = lambda: iisignature.sig(path, m)

            # Warmup
            _ = func()

            t_sec, peak_bytes = time_and_peak_memory(func, repeats=repeats)
            t_ms = t_sec * 1000.0
            alloc_kib = peak_bytes / 1024.0

            return {
                "N": N,
                "d": d,
                "m": m,
                "path_kind": path_kind,
                "operation": operation,
                "language": "python",
                "library": "iisignature",
                "method": "sig",
                "path_type": "ndarray",
                "t_ms": t_ms,
                "alloc_KiB": alloc_kib,
            }
        except Exception as e:
            print(f"iisignature signature benchmark failed: {e}", file=sys.stderr)
            return None

    elif operation == "logsignature":
        # iisignature logsig support
        if d < 2:
            # iisignature sometimes doesn't like d=1 for logsig
            return None
        try:
            path = make_path(d, N, path_kind)
            path = np.ascontiguousarray(path, dtype=np.float64)

            basis = iisignature.prepare(d, m, logsig_method)

            def func():
                iisignature.logsig(path, basis, logsig_method)

            # Warmup
            _ = iisignature.logsig(path, basis, logsig_method)

            t_sec, peak_bytes = time_and_peak_memory(func, repeats=repeats)
            t_ms = t_sec * 1000.0
            alloc_kib = peak_bytes / 1024.0

            return {
                "N": N,
                "d": d,
                "m": m,
                "path_kind": path_kind,
                "operation": operation,
                "language": "python",
                "library": "iisignature",
                "method": f"logsig({logsig_method})",
                "path_type": "ndarray",
                "t_ms": t_ms,
                "alloc_KiB": alloc_kib,
            }
        except Exception as e:
            print(f"iisignature logsignature benchmark failed: {e}", file=sys.stderr)
            return None

    # sigdiff not supported
    return None


# ------------------------------------------------------------------
# Main sweep + CSV writer
# ------------------------------------------------------------------

def run_bench() -> Path:
    cfg = load_config()
    Ns = cfg["Ns"]
    Ds = cfg["Ds"]
    Ms = cfg["Ms"]
    path_kind = cfg["path_kind"]
    repeats = int(cfg["repeats"])
    logsig_method = cfg["logsig_method"]
    operations = cfg["operations"]
    runs_dir = cfg["runs_dir"]
    libraries_cfg = cfg.get("libraries", [])

    print("=" * 60)
    print("Python Benchmark Suite")
    print("=" * 60)
    print("Configuration:")
    print(f"  path_kind     = {path_kind}")
    print(f"  Ns            = {Ns}")
    print(f"  Ds            = {Ds}")
    print(f"  Ms            = {Ms}")
    print(f"  operations    = {operations}")
    print(f"  repeats       = {repeats}")
    print(f"  logsig_method = {logsig_method} (iisignature; chen uses prepare_logsig)")
    print(f"  libraries     = {libraries_cfg}")
    print()

    # Setup libraries based on config
    setup_libraries(libraries_cfg)

    # Check if orchestrator overrides output CSV location
    env_out_csv = os.environ.get("BENCHMARK_OUT_CSV", "")
    if env_out_csv:
        csv_path = Path(env_out_csv)
        csv_path.parent.mkdir(parents=True, exist_ok=True)
        print(f"Orchestrator mode: writing results to {csv_path}")
        run_dir = None
    else:
        # Standalone mode: create own run folder
        from common import setup_run_folder, finalize_run_folder
        run_dir = setup_run_folder("benchmark_python", cfg)
        csv_path = run_dir / "results.csv"

    results: List[Dict[str, Any]] = []

    for N in Ns:
        for d in Ds:
            for m in Ms:
                for op in operations:
                    # chen-signatures
                    if "chen-signatures" in libraries_cfg:
                        res = bench_chen_signatures(d, m, N, path_kind, op, repeats=repeats)
                        if res is not None:
                            results.append(res)

                    # pysiglib
                    if "pysiglib" in libraries_cfg:
                        res = bench_pysiglib(d, m, N, path_kind, op, repeats=repeats)
                        if res is not None:
                            results.append(res)

                    # iisignature
                    if "iisignature" in libraries_cfg:
                        res = bench_iisignature(
                            d, m, N, path_kind, op, logsig_method, repeats=repeats
                        )
                        if res is not None:
                            results.append(res)

    fieldnames = [
        "N",
        "d",
        "m",
        "path_kind",
        "operation",
        "language",
        "library",
        "method",
        "path_type",
        "t_ms",
        "alloc_KiB",
    ]

    with csv_path.open("w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=fieldnames)
        writer.writeheader()
        for row in results:
            writer.writerow(row)

    print("=" * 60)
    print(f"Results written to: {csv_path}")

    # Only write summary in standalone mode
    if run_dir is not None:
        from common import finalize_run_folder
        summary = {
            "total_benchmarks": len(results),
            "libraries": sorted(list(set(r["library"] for r in results))),
            "operations": sorted(list(set(r["operation"] for r in results))),
            "output_csv": str(csv_path.name),
        }
        finalize_run_folder(run_dir, summary)

    return csv_path


if __name__ == "__main__":
    run_bench()

File: benchmark.jl
# benchmark.jl

using Revise
using BenchmarkTools
using StaticArrays
using ChenSignatures
using Printf
using Dates
using DelimitedFiles
using YAML

# -------- config loading --------

function load_config()
    config_path = joinpath(@__DIR__, "benchmark_config.yaml")
    @assert isfile(config_path) "Config file not found: $config_path"

    cfg = YAML.load_file(config_path)

    Ns       = get(cfg, "Ns", [150, 1000, 2000])
    Ds       = get(cfg, "Ds", [2, 6, 7, 8])
    Ms       = get(cfg, "Ms", [4, 6])
    path_str = get(cfg, "path_kind", "linear")
    runs_dir = get(cfg, "runs_dir", "runs")
    repeats  = get(cfg, "repeats", 5)
    operations_raw = get(cfg, "operations", ["signature", "logsignature"])

    path_kind = Symbol(path_str)
    operations = Symbol.(operations_raw)

    return (Ns = Ns, Ds = Ds, Ms = Ms,
            path_kind = path_kind, runs_dir = runs_dir,
            repeats = repeats, operations = operations)
end

# -------- path generators --------

# linear: [t, 2t, 2t, ...]
function make_path_linear_svec(d::Int, N::Int)
    ts = range(0.0, stop=1.0, length=N)
    [SVector{d,Float64}(ntuple(i -> (i == 1 ? t : 2t), d)) for t in ts]
end

function make_path_linear_matrix(d::Int, N::Int)
    ts = range(0.0, stop=1.0, length=N)
    path = Matrix{Float64}(undef, N, d)
    path[:, 1] .= ts
    for j in 2:d
        path[:, j] .= 2 .* ts
    end
    return path
end

# sinusoid: [sin(2π·1·t), sin(2π·2·t), ...]
function make_path_sin_svec(d::Int, N::Int)
    ts = range(0.0, stop=1.0, length=N)
    ω = 2π
    [SVector{d,Float64}(ntuple(i -> sin(ω * i * t), d)) for t in ts]
end

function make_path_sin_matrix(d::Int, N::Int)
    ts = range(0.0, stop=1.0, length=N)
    ω = 2π
    path = Matrix{Float64}(undef, N, d)
    for j in 1:d
        path[:, j] .= sin.(ω * j .* ts)
    end
    return path
end

function make_path(d::Int, N::Int, kind::Symbol, path_type::Symbol)
    if kind === :linear
        return path_type === :Matrix ? make_path_linear_matrix(d, N) : make_path_linear_svec(d, N)
    elseif kind === :sin
        return path_type === :Matrix ? make_path_sin_matrix(d, N) : make_path_sin_svec(d, N)
    else
        error("Unknown path_kind: $kind (expected :linear or :sin)")
    end
end

# -------- one benchmark case (Julia only) --------

function bench_case(d::Int, m::Int, N::Int, path_kind::Symbol, op::Symbol, path_type::Symbol, repeats::Int)
    path = make_path(d, N, path_kind, path_type)
    tensor_type = ChenSignatures.Tensor{Float64}

    # Determine path type string for output
    path_type_str = path_type === :Matrix ? "Matrix" : "Vector{SVector}"

    # Helper closures for the two operations
    run_sig() = signature_path(tensor_type, path, m)
    run_logsig() = ChenSignatures.log(signature_path(tensor_type, path, m))

    # Select function and warmup
    if op === :signature
        method_name = "signature_path"
        run_sig()
        # Note: We must interpolate local functions with $ for BenchmarkTools
        t_jl = @belapsed $run_sig() evals=1 samples=repeats
        a_jl = @allocated run_sig()
    elseif op === :logsignature
        method_name = "log"
        run_logsig()
        t_jl = @belapsed $run_logsig() evals=1 samples=repeats
        a_jl = @allocated run_logsig()
    else
        error("Unknown operation: $op")
    end

    t_ms      = t_jl * 1000
    alloc_KiB = a_jl / 1024

    return (N = N,
            d = d,
            m = m,
            path_kind = path_kind,
            operation = op,
            language = "julia",
            library = "ChenSignatures.jl",
            method = method_name,
            path_type = path_type_str,
            t_ms = t_ms,
            alloc_KiB = alloc_KiB)
end

# -------- sweep + write grid to file --------

function setup_run_folder(run_type::String, runs_root::String)
    ts = Dates.format(Dates.now(), "yyyymmdd_HHMMSS")
    run_dir = joinpath(runs_root, "$(run_type)_$(ts)")
    mkpath(run_dir)
    
    println("Created run folder: $run_dir")
    return run_dir, ts
end

function write_metadata(run_dir::String, run_type::String, ts::String, cfg)
    # Write config snapshot
    config_src = joinpath(@__DIR__, "benchmark_config.yaml")
    if isfile(config_src)
        cp(config_src, joinpath(run_dir, "benchmark_config.yaml"); force=true)
    end
    
    # Write resolved config
    config_resolved = joinpath(run_dir, "config_resolved.txt")
    open(config_resolved, "w") do f
        println(f, "Resolved Configuration")
        println(f, "=" ^ 60)
        println(f, "Ns:         $(cfg.Ns)")
        println(f, "Ds:         $(cfg.Ds)")
        println(f, "Ms:         $(cfg.Ms)")
        println(f, "path_kind:  $(cfg.path_kind)")
        println(f, "operations: $(cfg.operations)")
        println(f, "repeats:    $(cfg.repeats)")
    end
    
    # Write metadata
    metadata_file = joinpath(run_dir, "run_metadata.txt")
    open(metadata_file, "w") do f
        println(f, "Run Type: $run_type")
        println(f, "Timestamp: $ts")
        println(f, "Start Time: $(Dates.format(Dates.now(), "yyyy-mm-dd HH:MM:SS"))")
    end
    
    return metadata_file
end

function finalize_run_folder(run_dir::String, summary::Dict)
    summary_file = joinpath(run_dir, "SUMMARY.txt")
    open(summary_file, "w") do f
        println(f, "Benchmark Run Summary")
        println(f, "=" ^ 60)
        println(f)
        println(f, "Results:")
        println(f, "-" ^ 60)
        for (key, value) in summary
            println(f, "$key: $value")
        end
        println(f)
        println(f, "End Time: $(Dates.format(Dates.now(), "yyyy-mm-dd HH:MM:SS"))")
    end
    
    println("=" ^ 60)
    println("Run completed. Summary written to: $summary_file")
end

function run_bench()
    cfg = load_config()
    Ns, Ds, Ms = cfg.Ns, cfg.Ds, cfg.Ms
    path_kind  = cfg.path_kind
    runs_root  = joinpath(@__DIR__, cfg.runs_dir)
    repeats    = cfg.repeats
    operations = cfg.operations

    # Benchmark both path types
    path_types = [:Matrix, :VectorSVector]

    println("=" ^ 60)
    println("Julia Benchmark Suite")
    println("=" ^ 60)
    println("Configuration:")
    println("  path_kind  = $path_kind")
    println("  Ns         = $(Ns)")
    println("  Ds         = $(Ds)")
    println("  Ms         = $(Ms)")
    println("  operations = $(operations)")
    println("  path_types = $(path_types)")
    println("  repeats    = $repeats")
    println()

    # Check if orchestrator is overriding output location
    custom_csv = get(ENV, "BENCHMARK_OUT_CSV", "")
    if !isempty(custom_csv)
        # Orchestrator mode: write to specified location
        run_dir = dirname(custom_csv)
        file = custom_csv
        println("Orchestrator mode: writing to $file")
    else
        # Standalone mode: create own run folder
        run_dir, ts = setup_run_folder("benchmark_julia", runs_root)
        write_metadata(run_dir, "benchmark_julia", ts, cfg)
        file = joinpath(run_dir, "results.csv")
    end

    results = NamedTuple[]

    for N in Ns, d in Ds, m in Ms, op in operations, ptype in path_types
        if op === :sigdiff
            @warn "Julia benchmark: sigdiff not supported (skipping)" N d m path_kind
            continue
        end
        push!(results, bench_case(d, m, N, path_kind, op, ptype, repeats))
    end

    # Write results
    header = ["N", "d", "m", "path_kind", "operation", "language", "library", "method", "path_type", "t_ms", "alloc_KiB"]
    data = Array{Any}(undef, length(results) + 1, length(header))
    data[1, :] = header

    for (i, r) in enumerate(results)
        data[i + 1, 1] = r.N
        data[i + 1, 2] = r.d
        data[i + 1, 3] = r.m
        data[i + 1, 4] = String(r.path_kind)
        data[i + 1, 5] = String(r.operation)
        data[i + 1, 6] = r.language
        data[i + 1, 7] = r.library
        data[i + 1, 8] = r.method
        data[i + 1, 9] = r.path_type
        data[i + 1, 10] = r.t_ms
        data[i + 1, 11] = r.alloc_KiB
    end

    writedlm(file, data, ',')

    println("=" ^ 60)
    println("Results written to: $file")
    
    # Only write summary in standalone mode
    if isempty(custom_csv)
        summary = Dict(
            "total_benchmarks" => length(results),
            "path_types" => join(unique([r.path_type for r in results]), ", "),
            "operations" => join(unique([String(r.operation) for r in results]), ", "),
            "output_csv" => basename(file)
        )
        
        finalize_run_folder(run_dir, summary)
    end
    
    return file
end

run_bench()
